### PR TITLE
C++: fix equality refinement in new range analysis

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

**Comment by MathiasVP** [comment]
This also includes all of the previous PRs (https://github.com/github/codeql/pull/13045, https://github.com/github/codeql/pull/13116, https://github.com/github/codeql/pull/13117). Could we leave those out of this PR so that we can get this one merged faster?

File: cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisStage.qll [hunk 1]
```diff
@@ -729,7 +729,7 @@ module RangeStage<
   ) {
     exists(SemExpr e, D::Delta d1, D::Delta d2 |
       unequalFlowStepIntegralSsa(v, pos, e, d1, reason) and
-      boundedUpper(e, b, d1) and
+      boundedUpper(e, b, d2) and
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisStage.qll:732]: Would it be possible to add a range analysis library-test for this as well? That would also help make this commit independent of the other commits.

File: cpp/ql/test/experimental/query-tests/Security/CWE/CWE-193/constant-size/test.cpp [hunk 2]
```diff
@@ -78,3 +78,36 @@ void testInterproc(BigArray *arr) {
...
+        }
+
+        if (i == MAX_SIZE || n != arr[i]) {
+            if (i == MAX_SIZE) {
+                break;
+            }
+            n = arr[i];
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/test/experimental/query-tests/Security/CWE/CWE-193/constant-size/test.cpp:113]: cc @rdmarsh2

[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/test/experimental/query-tests/Security/CWE/CWE-193/constant-size/test.cpp]: Added

**Review by MathiasVP — DISMISSED** [review]
LGTM once DCA is happy!

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**